cartridge/failover.lua

```diff
@@ -200,65 +251,53 @@ local function failover_loop(args)
     checks({
         get_appointments = 'function',
     })
-    local confapplier = require('cartridge.confapplier')
-    local all_roles = require('cartridge.roles').get_all_roles()
 
-    ::start_over::
-
-    while pcall(fiber.testcancel) do
+    while true do
+        -- WARNING: implicit yield
         local appointments, err = FailoverError:pcall(args.get_appointments)
-        if appointments == nil then
-            -- don't log an error in case the fiber was cancelled
-            fiber.testcancel()
+        fiber.testcancel()
 
+        local csw1 = fiber.info()[fiber.id()].csw
+
+        if appointments == nil then
             log.warn('%s', err.err)
             vars.failover_err = FailoverError:new(
                 "Error fetching appointments: %s", err.err
             )
-            goto start_over
+            goto continue
         end
 
         vars.failover_err = nil
 
         if not accept_appointments(appointments) then
             -- nothing changed
-            goto start_over
+            goto continue
         end
 
-        -- The event may arrive during two-phase commit is in progress.
-        -- We should wait for the appropriate state.
-        local state = confapplier.wish_state('RolesConfigured', math.huge)
-        if state ~= 'RolesConfigured' then
-            log.info('Skipping failover step - state is %s', state)
-            goto start_over
-        end
-
-        log.info('Failover triggered')
-        confapplier.set_state('ConfiguringRoles')
-
-        local ok, err = FailoverError:pcall(function()
-            box.cfg({
-                read_only = not vars.cache.is_rw,
-            })
-
-            for _, role_name in ipairs(all_roles) do
-                local mod = service_registry.get(role_name)
-                local _, err = apply_config(mod)
-                if err then
-                    log.error('Role %q failover failed', mod.role_name)
-                    log.error('%s', err)
-                end
+        -- Cancel all pending tasks
+        for id, task in pairs(vars.schedule) do
+            if task:status() == 'dead' then
+                vars.schedule[id] = nil
+            elseif not task.storage.is_busy then
+                vars.schedule[id] = nil
+                task:cancel()
+            -- else
+                -- preserve busy tasks
             end
+        end
 
-            return true
-        end)
-
-        if ok then
-            log.info('Failover step finished')
-        else
-            log.warn('Failover step failed: %s', err)
+        -- Schedule new task
+        do
+            local task = fiber.new(reconfigure_all, vars.cache.active_leaders)
+            local id = task:id()
+            task:name('cartridge.failover.task')
+            vars.schedule[id] = task
```

Comment:
Is order important here?

I assume that only last task should win, but seems it strictly depends on tarantool fiber scheduler.

Reply:
There is always at most 1 busy task, and at most 1 pending.
Scheduling doesn't affect execution priority.

Reply:
Order isn't important.
The only reason to use `schedule[id] = task` is that it looks prettier than `schedule[task] = true` (IMO).